### PR TITLE
VS-142: Extended Analyzer Functionality to support EF-based Queries

### PR DESCRIPTION
**Pull request by Ravi-Raghavan**
This PR has extended the functionality of the MongoDB Analyzer to render analysis for EF-based Queries.

### CODE REVIEW

**Review by BorisDog — CHANGES_REQUESTED** [review]

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 1]
```diff
@@ -0,0 +1,147 @@
...
+        var telemetry = AnalysisUtilities.GetTelemetry(EFAnalysis, stats, sw);
+        if (telemetry.ExpressionsFound > 0)
+        {
+            context.Telemetry.EFAnalysisResult(telemetry);
+        }
+
+        return telemetry.ExpressionsFound > 0;
+    }
+
+    private static bool IsDriverOrLinqException(MQLResult mqlResult)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:55]: Probably need to refactor this out

File: src/MongoDB.Analyzer/DiagnosticsRules.cs [hunk 2]
```diff
@@ -25,6 +25,11 @@ internal static class DiagnosticsRules
     private const string DiagnosticIdBuilders2MQL = "MABuilders1001";
     private const string DiagnosticIdNotSupportedBuilderExpression = "MABuilders2001";
 
+    //EF
+    private const string DiagnosticIdEF2MQL = "MAEF1001";
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/DiagnosticsRules.cs:29]: Needs also to update `AnalyzerReleases.Shipped.md`

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        var db = new MyDbContext(dbContextOptions.Options);
+        var users_query = db.Users.OrderBy(u => u.Age).ThenBy(u => u.Height);
+        var users_query2 = db.Users.OrderBy(u => u.Age).ThenByDescending(u => u.Height);
+    }
+}
+
+internal class MyDbContext : DbContext
+{
+    public DbSet<User> Users { get; set; }
+
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs:85]: No need for empty line

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs [hunk 4]
```diff
@@ -18,7 +18,7 @@ namespace MongoDB.Analyzer.Core.Linq;
 
 internal static class LinqExpressionProcessor
 {
-    public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalysisContext context)
+    public static ExpressionsAnalysis ProcessSemanticModel(MongoAnalysisContext context, AnalysisType analysisType = AnalysisType.Linq)
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:21]: Validate `analysisType` value.

**Review by BorisDog — CHANGES_REQUESTED** [review]
Look really good, few comments.

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs (hunk 3 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs:85]: Done

File: src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs (hunk 4 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/LinqExpressionProcessor.cs:21]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs (hunk 1 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:55]: Done

File: src/MongoDB.Analyzer/DiagnosticsRules.cs (hunk 2 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/DiagnosticsRules.cs:29]: Done

File: src/MongoDB.Analyzer.Helpers/MongoDB.Analyzer.Helpers.csproj [hunk 5]
```diff
@@ -24,4 +24,10 @@
     <PackageReference Include="MongoDB.Driver" Version="2.19.0" Condition="'$(Configuration)' == 'DRIVER_2_19'" />
     <PackageReference Include="MongoDB.Driver" Version="2.21.0" Condition="'$(Configuration)' == 'DRIVER_2_21'" />
   </ItemGroup>
+  <ItemGroup>
+    <None Remove="EF\" />
+  </ItemGroup>
+  <ItemGroup>
+    <Folder Include="EF\" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer.Helpers/MongoDB.Analyzer.Helpers.csproj]: Done

File: tests/MongoDB.Analyzer.Tests/MongoDB.Analyzer.Tests.csproj [hunk 6]
```diff
@@ -23,12 +23,20 @@
...
   <ItemGroup>
     <ProjectReference Include="..\..\src\MongoDB.Analyzer\MongoDB.Analyzer.csproj" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="EF\" />
+  </ItemGroup>
+  <ItemGroup>
+    <Folder Include="EF\" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/MongoDB.Analyzer.Tests.csproj]: Done

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DriverVersionHelper.cs [hunk 7]
```diff
@@ -25,17 +25,17 @@ internal static class DriverVersionHelper
...
-        //   NuGetVersion.Parse("2.25.0")
-        // }
+        DriverVersions = new[]
+        {
+           NuGetVersion.Parse("2.12.4"),
+           NuGetVersion.Parse("2.18.0"),
+           NuGetVersion.Parse("2.19.0"),
+           NuGetVersion.Parse("2.21.0"),
+           NuGetVersion.Parse("2.25.0")
+         };
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/Infrastructure/DriverVersionHelper.cs]: Done

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs [hunk 8]
```diff
@@ -0,0 +1,96 @@
...
+using Microsoft.EntityFrameworkCore;
+using MongoDB.Analyzer.Tests.Common.DataModel;
+
+namespace MongoDB.Analyzer.Tests.Common.TestCases.EF;
+
+public sealed class EFBasic
+{
+    [MQLEF("aggregate([{ \"$match\" : { \"Name\" : \"Bob\", \"Age\" : { \"$gt\" : 16, \"$lte\" : 21 } } }])")]
+    [MQLEF("aggregate([{ \"$match\" : { \"Name\" : \"Bob\", \"CustomerId\" : 21 } }])")]
+    public void EF_Where_Queries()
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs]: Done

File: src/MongoDB.Analyzer/MongoDB.Analyzer.csproj [hunk 9]
```diff
@@ -33,4 +33,10 @@
     <PackageReference Include="System.Runtime.Loader" Version="4.3.0" />
   </ItemGroup>
 
+  <ItemGroup>
+    <None Remove="Core\EF\" />
+  </ItemGroup>
+  <ItemGroup>
+    <Folder Include="Core\EF\" />
+  </ItemGroup>
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/MongoDB.Analyzer.csproj]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 10]
```diff
@@ -0,0 +1,140 @@
...
+                semanticContext.ReportDiagnostics(diagnosticDescriptor, decoratedMessage, locations);
+                mqlCount++;
+            }
+            else if (mqlResult.Exception != null)
+            {
+                var isDriverOrLinqException = AnalysisUtilities.IsDriverOrLinqException(mqlResult);
+
+                if (isDriverOrLinqException || settings.OutputInternalExceptions)
+                {
+                    var diagnosticDescriptor = DiagnosticsRules.DiagnosticRuleNotSupportedLinqExpression;
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:119]: EF rule

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+
+        foreach (var analysisContext in EFExpressionAnalysis.AnalysisNodeContexts)
+        {
+            var mqlResult = compilationResult.LinqTestCodeExecutor.GenerateMql(analysisContext.EvaluationMethodName);
+            var locations = analysisContext.Node.Locations;
+
+            if (mqlResult.Mql != null)
+            {
+                var mql = analysisContext.Node.ConstantsRemapper.RemapConstants(mqlResult.Mql);
+                var diagnosticDescriptor = mqlResult.Linq3Only ? DiagnosticsRules.DiagnosticRuleNotSupportedEFExpression : DiagnosticsRules.DiagnosticRuleEF2MQL;
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:108]: No linq2 in ef I think. Verify with Damien?

**Review by BorisDog — CHANGES_REQUESTED** [review]
Some comments

File: tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs [hunk 12]
```diff
@@ -42,8 +42,11 @@ internal static class DiagnosticsAnalyzer
 #endif
 
         var packages = ImmutableArray.Create(new PackageIdentity("MongoDB.Driver", driverVersion));
+        var EFPackage = ImmutableArray.Create(new PackageIdentity("Microsoft.EntityFrameworkCore", "3.1.0"));
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests/Infrastructure/DiagnosticsAnalyzer.cs:45]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs (hunk 10 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:119]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+                mqlCount++;
+            }
+            else if (mqlResult.Exception != null)
+            {
+                var isDriverOrLinqException = AnalysisUtilities.IsDriverOrLinqException(mqlResult);
+
+                if (isDriverOrLinqException || settings.OutputInternalExceptions)
+                {
+                    var diagnosticDescriptor = DiagnosticsRules.DiagnosticRuleNotSupportedLinqExpression;
+                    var message = AnalysisUtilities.GetExceptionMessage(mqlResult.Exception, typesMapper, AnalysisType.Linq);
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:120]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+            var diagnostics = Diagnostic.Create(
+                DiagnosticsRules.DiagnosticRuleNotSupportedLinqExpression,
+                invalidLinqNode.OriginalExpression.GetLocation(),
+                AnalysisUtilities.DecorateMessage(invalidLinqNode.Errors.FirstOrDefault(), driverVersionString, context.Settings));
+
+            semanticContext.ReportDiagnostic(diagnostics);
+        }
+    }
+
+    private static AnalysisStats ReportMqlOrInvalidExpressions(MongoAnalysisContext context, ExpressionsAnalysis EFExpressionAnalysis)
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:78]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+            return;
+        }
+
+        var driverVersion = ReferencesProvider.GetMongoDBDriverVersion(semanticContext.SemanticModel.Compilation.References);
+        var driverVersionString = driverVersion?.ToString(3);
+
+        foreach (var invalidLinqNode in EFExpressionAnalysis.InvalidExpressionNodes)
+        {
+            var diagnostics = Diagnostic.Create(
+                DiagnosticsRules.DiagnosticRuleNotSupportedLinqExpression,
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:70]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+        var telemetry = AnalysisUtilities.GetTelemetry(EFAnalysis, stats, sw);
+        if (telemetry.ExpressionsFound > 0)
+        {
+            context.Telemetry.EFAnalysisResult(telemetry);
+        }
+
+        return telemetry.ExpressionsFound > 0;
+    }
+
+    private static void ReportInvalidExpressions(MongoAnalysisContext context, ExpressionsAnalysis EFExpressionAnalysis)
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:55]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs [hunk 17]
```diff
@@ -0,0 +1,140 @@
...
+
+namespace MongoDB.Analyzer.Core.Linq;
+
+internal static class EFAnalyzer
+{
+    public static bool AnalyzeEFQueryable(MongoAnalysisContext context)
+    {
+        var sw = Stopwatch.StartNew();
+        var stats = AnalysisStats.Empty;
+        ExpressionsAnalysis EFAnalysis = null;
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:26]: Done

File: src/MongoDB.Analyzer/Core/Utilities/SymbolExtensions.cs [hunk 18]
```diff
@@ -168,6 +169,9 @@ public static bool IsDefinedInMongoLinqOrSystemLinq(this ISymbol symbol)
             symbol?.ContainingAssembly.Name == AssemblyMongoDBDriver;
     }
 
+    public static bool IsEF(this ITypeSymbol typeSymbol) =>
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/SymbolExtensions.cs:172]: Done

File: src/MongoDB.Analyzer/Core/Utilities/AnalysisUtilities.cs [hunk 19]
```diff
@@ -26,6 +26,13 @@ internal static class AnalysisUtilities
             _ => typesMapper.RemapTypes(exception.InnerException?.Message)
         };
 
+    public static bool IsDriverOrLinqException(Linq.MQLResult mqlResult)
+    {
+        var source = mqlResult.Exception.InnerException?.Source;
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Utilities/AnalysisUtilities.cs:31]: Done

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+    }
+
+    [MQLEF("aggregate([{ \"$match\" : { \"Name\" : \"Bob\", \"Age\" : { \"$gt\" : 16, \"$lte\" : 21 } } }])")]
+    [MQLEF("aggregate([{ \"$match\" : { \"Name\" : \"Bob\", \"CustomerId\" : 21 } }])")]
+    public void Where()
+    {
+        var dbContextOptions = new DbContextOptionsBuilder<MyDbContext>();
+        var db = new MyDbContext(dbContextOptions.Options);
+        var users_query = db.Users.Where(u => u.Name == "Bob" && u.Age > 16 && u.Age <= 21);
+        var customers_query = db.Customers.Where(c => c.Name == "Bob" & c.CustomerId == 21);
```
[the remark this answers is not in the document]
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs:78]: Done

File: src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs (hunk 11 above)
Comment (Ravi-Raghavan) [reply src/MongoDB.Analyzer/Core/Linq/EFAnalyzer.cs:108]: Only LINQ3

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs [hunk 21]
```diff
@@ -0,0 +1,122 @@
...
+namespace MongoDB.Analyzer.Tests.Common.TestCases.EF;
+
+public sealed class EFBasic
+{
+    public MyDbContext GetDbContext() => new MyDbContext(new DbContextOptionsBuilder<MyDbContext>().Options);
+    public DbContextOptions GetDbContextOptions() => new DbContextOptionsBuilder<MyDbContext>().Options;
+    public DbSet<Customer> GetDbSet_Customers() => new MyDbContext(new DbContextOptionsBuilder<MyDbContext>().Options).Customers;
+    public DbSet<User> GetDbSet_Users() => new MyDbContext(new DbContextOptionsBuilder<MyDbContext>().Options).Users;
+
+
```
Comment (BorisDog) [inline tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs:29]: Remove empty line, and place those helper method at the bottom of the file.

File: src/MongoDB.Analyzer/Core/Utilities/SymbolExtensions.cs [hunk 22]
```diff
@@ -155,6 +156,9 @@ symbol switch
         _ => symbol.IsContainedInLambda(parentNode)
     };
 
+    public static bool IsDBSet(this ITypeSymbol typeSymbol) =>
+        typeSymbol?.ContainingNamespace?.ToDisplayString() == NamespaceEF;
```
Comment (BorisDog) [inline src/MongoDB.Analyzer/Core/Utilities/SymbolExtensions.cs:160]: Sorry, I meant let's check for that it's a symbol of DBSet type, not just a type defined in NamespaceEF.

File: tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs (hunk 21 above)
Comment (Ravi-Raghavan) [reply tests/MongoDB.Analyzer.Tests.Common.TestCases/EF/EFBasic.cs]: Done

**Review by BorisDog — APPROVED** [review]
LGTM!

**Pull request merged**